basic_pipeline/3_Source.md

```diff
@@ -0,0 +1,132 @@
+The callbacks we are about to implement will be called once the transition between playback states occurs.
+
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ @impl true
+ def handle_stopped_to_prepared(_ctx, state) do
+  raw_file_binary = File.read!(state.location)
+  content = String.split(raw_file_binary, "\n")
+  state = %{state | content: content}
+  {{:ok, [caps: {:output, %Basic.Formats.Packet{type: :custom_packets}}]}, state}
+ end
+
+ @impl true
+ def handle_prepared_to_stopped(_ctx, state) do
+  state = %{state | content: nil}
+  {:ok, state}
+ end
+ ...
+end
+
+```
+In the case of the first callback, `handle_stopped_to_prepared/2`, what we do is that we are reading the file from the location specified in the options structure (which we have saved in the state of the element).
+Then we split the content of the file to get the particular packets and save the list of those packets in the state of the element.
+An interesting thing here is the action we are returning - the `:caps` action. That means that we want to transmit the information about the supported caps through the `output` pad, to the next element in the pipeline. In the [9th chapter](9_Caps.md) you will find out more about caps and formats and learn why it is required to do so.
+The second callback, `handle_prepared_to_stopped`, defines the behavior of the Source element while we are stopping the pipeline. What we want to do is to clear the content buffer in the state of our element.
+# Demands
+Before going any further let's stop for a moment and talk about the demands. Do you remember, that the `:output` pad is working in the pulling mode? That means that the succeeding element have to ask the Source element for the data to be sent and our element has to take care of keeping that data in some kind of buffer until it is requested. 
+Once the succeeding element requests for the data, the `handle_demand/4` callback will be invoked - therefore it would be good for us to define it:
+```Elixir
+#FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ @impl true
+ def handle_demand(:output, 0, :buffers, _ctx, state) do
+  {:ok, state}
+ end
+
+ @impl true
+ def handle_demand(:output, size, :buffers, _ctx, state) do
+  if state.content == [] do
+    {{:ok, end_of_stream: :output}, state}
+  else
+    [chosen | rest] = state.content
+    state = %{state | content: rest}
+    action = [buffer: {:output, %Buffer{payload: chosen}}]
+    action = if size > 1, do: action ++ [redemand: :output], else: action
```

Comment:
this `if` is redundant, redemnad will cause execution of `handle_demand`, but in our case handle_demand with size 0 is noop, so I wouldn't bother with this additional `if`.

Reply:
indeed.
I have changed it both here and in the code.
done